models/shopcart.py

```diff
@@ -18,16 +18,23 @@ def __init__(self, uid=0, products=None):
         :param products: dict of products <products id, quantity of product>
         """
         self.uid = int(uid)
-        self.products = self.__validate_products(products)
-        
+        try:
+            self.products = self.__validate_products(products)
+        except DataValidationError as e:
+            print('ERROR: Data Validation error' +e.args[0])           
+
     def save(self):
         """ Saves a Shopcart in the database """
         Shopcart.__data.append(self)
 
     def add_product(self, pid, quant=1):
         """ Adds a tuple of product, quantity to the product dict """
         pq_tup = (pid,quant)
-        self.products.update( self.__validate_products(pq_tup) )
+        try:
+            self.products.update( self.__validate_products(pq_tup) )
+        except DataValidationError as e:
```

Comment:
So the idea is that there's not error catching at all in `products.py`. If something on the database side raises an error, let it be thrown to the caller, and by caller I mean the API. It's better to throw the error up the call chain to the API so it can construct a proper response for the user.